### PR TITLE
new feature FRAGMENT_FROM_FILE

### PR DESCRIPTION
**Pull request by BenGardiner**
proposed in #604. This introduces the ability to include 'fragments' of strictdoc files. Dual-purposing `[SECTION]` as the include target _and_ a section-as-usual didn't seem possible in the grammar and `processor.py` structure so a `[FRAGMENT]` top-level directive for include files is proposed here.

### CODE REVIEW

File: docs/sphinx/source/strictdoc.rst [hunk 1]
```diff
@@ -519,6 +519,42 @@ Special feature of ``[COMPOSITE_REQUIREMENT]``: like ``[SECTION]`` element, the
 often, a more basic combination of nested ``[SECTION]`` and ``[REQUIREMENT]``
 elements should do the job.
 
+Include Files
+~~~~~~~~~~~~~
+
+Strictdoc ``.sdoc`` files can be built-up from including other fragment documents.
```
Comment (stanislaw) [inline docs/sphinx/source/strictdoc.rst:525]: `StrictDoc`

File: docs/sphinx/source/strictdoc.rst [hunk 2]
```diff
@@ -519,6 +519,42 @@ Special feature of ``[COMPOSITE_REQUIREMENT]``: like ``[SECTION]`` element, the
...
+
+The ``[INCLUDE_SECTION_FROM_FILE]`` element can be used anywhere a ``[SECTION]`` can
```
Comment (stanislaw) [inline docs/sphinx/source/strictdoc.rst:527]: How about we call this `FRAGMENT_FROM_FILE`?

File: docs/sphinx/source/strictdoc.rst [hunk 3]
```diff
@@ -519,6 +519,42 @@ Special feature of ``[COMPOSITE_REQUIREMENT]``: like ``[SECTION]`` element, the
 often, a more basic combination of nested ``[SECTION]`` and ``[REQUIREMENT]``
 elements should do the job.
 
+Include Files
```
Comment (stanislaw) [inline docs/sphinx/source/strictdoc.rst:522]: `Include files` (the project follows European style for the capitalisation of headers)

File: docs/sphinx/source/strictdoc.rst [hunk 4]
```diff
@@ -519,6 +519,42 @@ Special feature of ``[COMPOSITE_REQUIREMENT]``: like ``[SECTION]`` element, the
...
+    TITLE: StrictDoc
+
+    [INCLUDE_SECTION_FROM_FILE]
+    FILE: include.ssec
+
+Then the referenced file, ``include.ssec``:
+
+.. code-block:: text
+
+    [FRAGMENT]
```
Comment (stanislaw) [inline docs/sphinx/source/strictdoc.rst:548]: In the https://github.com/strictdoc-project/strictdoc/pull/632, I have tried to work around the `Dual-purposing [SECTION] as the include target and a section-as-usual` problem that you mentioned by making it explicit what's a fragment (`FRAGMENT`) and its contents (the usual `SECTIONS`).

So the example below should have:

```
[FRAGMENT]

[SECTION]
TITLE: Free text

[/SECTION]
```

**Review by stanislaw — CHANGES_REQUESTED** [review]
Thanks a lot for creating this PR. I believe that I have pushed it further to the right direction here: https://github.com/strictdoc-project/strictdoc/pull/632. I didn't know how I would open a MR back to your repository, so I simply opened another one here.

I suggest that you rebase your PR on top of the latest `main` branch, cherry-pick my code review commit and address all (relatively minor) comments. After that, the work should be mergeable.

The resolution of the fragments might need some attention though - it was a result of a quick work by me and the code there could surely be improved. Please check on your side if everything still works like you expect it to work.

File: strictdoc/backend/sdoc/grammar/grammar.py [hunk 5]
```diff
@@ -94,7 +112,16 @@
 ;
 
 SectionOrRequirement[noskipws]:
-  '\n' (Section | Requirement | CompositeRequirement)
+  '\n' (Section | Requirement | CompositeRequirement | IncludeFromFile)
+;
+
+IncludeFromFile[noskipws]:
```
Comment (stanislaw) [inline strictdoc/backend/sdoc/grammar/grammar.py:118]: Let's call this `FragmentFromFile` and `[FRAGMENT_FROM_FILE]`.

File: strictdoc/backend/sdoc/grammar/grammar.py [hunk 6]
```diff
@@ -94,7 +112,16 @@
...
+  '[INCLUDE_SECTION_FROM_FILE]' '\n'
+  ('UID: ' uid = /.+$/ '\n')?
```
Comment (stanislaw) [inline strictdoc/backend/sdoc/grammar/grammar.py:120]: Is there a use case for maintaining these three optional fields: `UID`, `LEVEL` and `TITLE`?

**Comment by BenGardiner** [comment]
> Thanks a lot for creating this PR. I believe that I have pushed it further to the right direction here: #632. I didn't know how I would open a MR back to your repository, so I simply opened another one here.

Apologies for that extra work. I'm not sure if I could have prepared the PR differently to enable that. But happy to learn how.
 
> I suggest that you rebase your PR on top of the latest `main` branch, cherry-pick my code review commit and address all (relatively minor) comments. After that, the work should be mergeable.

No problem -- I can manage all the git backflips necessary

**Comment by BenGardiner** [comment]
I've addressed all the review feedback here and in #632 .

**Review by stanislaw — APPROVED** [review]

**Comment by stanislaw** [comment]
Thanks for doing all this work. I felt like I was almost reviewing my own code 😄 

Good catch with the before and after parts when extending the section_contents 👍 

Are you planning to push anything on top of this change or should I release a new Pip?

**Comment by BenGardiner** [comment]
> Thanks for doing all this work. I felt like I was almost reviewing my own code 😄

It's my pleasure! Thank you and the team for the strictdoc project! 

> Are you planning to push anything on top of this change or should I release a new Pip?

Yes, _maybe_. Depending on discussion #630 ...

**Pull request merged**